feat(editor): Add v1 banner

ivov opened the pull request:
https://linear.app/n8n/issue/N8N-6485

ivov committed:
feat(editor): Add v1 banner

ivov committed:
fix: Export MySQL and PG migrations

ivov commented on packages/editor-ui/src/App.vue:
```diff
@@ -278,12 +281,12 @@ export default defineComponent({
 .header {
 	grid-area: header;
-	z-index: 999;
+	z-index: 99;
```
Decrease here to allow the v1 banner to display on top of header and sidebar, but without having to increase the banner's `z-index` so much that it would even display on top of the backdrop used for modals.

ivov committed:
fix: Do not throw on missing setting

ivov committed:
fix: Icon selection in callout

ivov committed:
test: Fix tests

ivov committed:
refactor: Add down migrations

ivov committed:
refactor: Spacing

ivov committed:
test: Fix another test

ivov committed:
refactor: Reuse type

ivov committed:
refactor: Specify banner-allowed views

ivov commented on packages/editor-ui/src/components/V1Banner.vue:
```diff
@@ -0,0 +1,76 @@
+const { dismissBanner } = useUIStore();
+
+const shouldDisplay = computed(() => {
+	if (!useRootStore().versionCli.startsWith('1.')) return false;
```
Disable L37 to test.

netroy commented on packages/cli/src/databases/migrations/mysqldb/1681134145998-AddV1BannerSetting.ts:
```diff
@@ -0,0 +1,16 @@
+
+export class AddV1BannerSetting1681134145998 implements ReversibleMigration {
+	async up({ queryRunner, tablePrefix }: MigrationContext) {
+		await queryRunner.query(
```
do we really want to setup a migration and stuff on the backend for this? could we not use `localStorage.set` and live with users seeing the banner again if they logout and then log back in?

That way we don't have to worry about this migration lingering around until v2.0 (or whenever we decide to flatten the migrations).

csuermann replied:
I'd advise against introducing local storage as yet another place for storing fragments of state. Using the DB for this is perfectly ok IMO, but I agree that we should aim to do it without DB migrations, as no schema adjustment is needed. Could we not simply treat the _absence_ of the `ui.banners.v1.dismissed` entry as `false`?

ivov committed:
refactor: Remove migrations

ivov committed:
refactor: Adjust endpoints

netroy commented on packages/editor-ui/src/components/V1Banner.vue:
```diff
@@ -0,0 +1,76 @@
+
+<style module lang="scss">
+.v1-banner {
+	position: absolute;
```
we could use `position: sticky` instead, or even better if we can fit this banner in the existing grid defined in `App.vue`.

ivov replied:
With `absolute` the banner already [remains in viewport on scroll](https://share.cleanshot.com/pWWg4N4x), which I think is what we want. See e.g. `/templates` 

With `sticky` the banner does not render - I'd need to look into the stacking context to debug, plus I'd rather keep it out of the flow of the document to minimize the chance of causing layout issues. So for MVP I think it's okay? Let me know otherwise.

netroy replied:
with `sticky` to work, you'd need to move `<V1Banner />` to line 3 in `App.vue`.

ivov replied:
That [blocks out scrolling](https://share.cleanshot.com/YsLLCrlB) in the templates feed.

ivov committed:
refactor: Create `SettingsRepository` method

ivov committed:
refactor: Rely on config instead of DB

ivov committed:
fix: Keep config and DB in sync

Otherwise, refreshing immediately after permanently dismissing causes banner to reappear.

netroy commented on packages/cli/src/controllers/owner.controller.ts:
```diff
@@ -148,4 +148,12 @@ export class OwnerController {
+	@Post('/dismiss-v1')
+	async dismissBanner() {
+		await this.settingsRepository.saveSetting('ui.banners.v1.dismissed', JSON.stringify(true));
+		this.config.set('ui.banners.v1.dismissed', true);
```
perhaps this should also be in `settingsRepository.saveSetting` (if `loadOnStartup` is `true`)

netroy commented on packages/cli/src/Server.ts:
```diff
@@ -402,7 +407,7 @@ export class Server extends AbstractServer {
 	 * Returns the current settings for the frontend
 	 */
-	getSettingsForFrontend(): IN8nUISettings {
+	async getSettingsForFrontend(): Promise<IN8nUISettings> {
```
this doesn't need to be async anymore

ivov committed:
refactor: Make `getSettingsForFrontend` sync again

ivov committed:
refactor: Move config fix to `SettingsRepository`

netroy reviewed (approved):
🙏🏽

netroy committed:
Merge remote-tracking branch 'origin/release-v1' into n8n-6485-v1-banner

janober commented:
Got released with [`n8n@1.0.5`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.0.5)
